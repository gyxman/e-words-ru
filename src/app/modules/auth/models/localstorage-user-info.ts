export type LocalstorageUserInfo = Readonly<{
    token: string;
    id: string;
}>;
