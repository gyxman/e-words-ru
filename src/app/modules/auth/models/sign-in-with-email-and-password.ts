export type SignInWithEmailAndPassword = Readonly<{
    email: string;
    password: string;
}>;
