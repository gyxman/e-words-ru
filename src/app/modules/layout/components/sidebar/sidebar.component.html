<ul class="list">
    <li
        *ngFor="let item of allMenuItems$ | async; trackBy: identityTrackByFunction"
        class="item"
    >
        <a class="link" [routerLink]="item.navigateTo">{{ item.title }}</a>
    </li>
</ul>
