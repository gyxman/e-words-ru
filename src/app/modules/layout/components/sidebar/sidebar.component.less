@import '../../../../../styles/index';

:host {
    display: block;
}

.list {
    padding: 10px 0;
    margin: 0;
    list-style: none;
}

.link {
    .font-h5();
    display: block;
    padding: 10px 20px;
    text-decoration: none;
    color: inherit;

    &:hover {
        background-color: @color-gray-extra-light;
    }
}
