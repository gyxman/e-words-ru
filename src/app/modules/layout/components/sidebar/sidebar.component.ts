import {Component, EventEmitter, HostListener, Output} from '@angular/core';
import {SidebarFacadeService} from '../../services/sidebar-facade.service';

@Component({
    selector: 'app-sidebar',
    templateUrl: './sidebar.component.html',
    styleUrls: ['./sidebar.component.less'],
})
export class SidebarComponent {
    @Output() sidebarOnClick = new EventEmitter();

    @HostListener('click') onClick() {
        this.sidebarOnClick.emit();
    }

    readonly identityTrackByFunction = (_index: number, item: any) => item;
    readonly allMenuItems$ = this.sidebarFacadeService.allMenu$;

    constructor(private sidebarFacadeService: SidebarFacadeService) {}
}
