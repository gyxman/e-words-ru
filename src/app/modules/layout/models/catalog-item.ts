import {ComponentEnum} from '../../exercises/enums/component.enum';

export type CatalogItem = Readonly<{
    name: string;
    component: ComponentEnum;
}>;
