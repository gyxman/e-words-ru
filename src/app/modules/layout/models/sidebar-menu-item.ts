export type SidebarMenuItem = Readonly<{
    title: string;
    key: string;
    navigateTo: string;
}>;
