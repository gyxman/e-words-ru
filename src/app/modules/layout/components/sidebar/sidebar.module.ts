import {NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {SidebarComponent} from './sidebar.component';
import {SidebarFacadeService} from '../../services/sidebar-facade.service';
import {RouterModule} from '@angular/router';

@NgModule({
    imports: [CommonModule, RouterModule],
    declarations: [SidebarComponent],
    exports: [SidebarComponent],
    providers: [SidebarFacadeService],
})
export class SidebarModule {}
