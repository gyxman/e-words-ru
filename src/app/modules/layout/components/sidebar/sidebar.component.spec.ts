import {ComponentFixture, TestBed} from '@angular/core/testing';
import {SidebarComponent} from './sidebar.component';
import {RouterTestingModule} from '@angular/router/testing';
import {SidebarFacadeService} from '../../services/sidebar-facade.service';
import {instance, mock, when} from 'ts-mockito';
import {BehaviorSubject} from 'rxjs';
import {SidebarMenuItem} from '../../models/sidebar-menu-item';
import {Component} from '@angular/core';
import {PageObject} from '../../../../testing/page-object';
import {LayoutRouteEnum} from '../../enums/layout-route.enum';

@Component({
    template:
        '<app-sidebar automation-id="sidebar" (sidebarOnClick)="onClick()"></app-sidebar>',
})
class TestComponent {
    onClick() {}
}

describe('SidebarComponent', () => {
    let testComponent: TestComponent;
    let fixture: ComponentFixture<TestComponent>;
    let pageObject: PageObject<TestComponent>;
    let sidebarFacadeServiceMock: SidebarFacadeService;
    let allMenu$: BehaviorSubject<Array<SidebarMenuItem>>;

    beforeEach(() => {
        sidebarFacadeServiceMock = mock(SidebarFacadeService);
        allMenu$ = new BehaviorSubject([]);
    });

    beforeEach(() => {
        when(sidebarFacadeServiceMock.allMenu$).thenReturn(allMenu$);
    });

    beforeEach(() => {
        TestBed.configureTestingModule({
            declarations: [TestComponent, SidebarComponent],
            imports: [RouterTestingModule],
            providers: [
                {
                    provide: SidebarFacadeService,
                    useFactory: () => instance(sidebarFacadeServiceMock),
                },
            ],
        }).compileComponents();
    });

    beforeEach(() => {
        fixture = TestBed.createComponent(TestComponent);
        testComponent = fixture.componentInstance;
        pageObject = new PageObject(fixture);
    });

    it('Если компонент загрузился, то все элементы отрисовались', () => {
        // arrange
        allMenu$.next([
            {
                title: 'Добавить слово',
                key: 'addWord',
                navigateTo: LayoutRouteEnum.AddWord,
            },
        ]);

        // act
        fixture.detectChanges();

        // assert
        expect(fixture.nativeElement).toMatchSnapshot();
    });

    it('Если происходит клик по компоненту левого меню, то сайдбар закрывается', () => {
        // arrange
        jest.spyOn(testComponent, 'onClick');

        fixture.detectChanges();

        // act
        pageObject.getByAutomationId('sidebar').triggerEventHandler('click', null);

        // assert
        expect(testComponent.onClick).toHaveBeenCalledTimes(1);
    });
});
