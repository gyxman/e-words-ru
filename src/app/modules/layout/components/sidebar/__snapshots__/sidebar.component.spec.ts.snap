// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarComponent Если компонент загрузился, то все элементы отрисовались 1`] = `
<div
  id="root0"
>
  <app-sidebar
    automation-id="sidebar"
  >
    <ul
      class="list"
    >
      
      <li
        class="item"
      >
        <a
          class="link"
          href="/add-word"
        >
          Добавить слово
        </a>
      </li>
    </ul>
  </app-sidebar>
</div>
`;
